### PR TITLE
[fix][build] Fix the pulsar-all image may use the wrong upstream image

### LINKED ISSUE #20420: [Bug] pulsar-all 3.0.0 docker image contains Pulsar 2.11.0
### Version

Ubuntu 20.04, Pulsar 3.0.0

### Minimal reproduce step

image

### What did you expect to see?

pulsar-related libs in pulsar-all:3.0.0 image should be 3.0.0 also.

### What did you see instead?

pulsar-related libs in pulsar-all:3.0.0 is 2.11.0.

### PR DESCRIPTION
**Pull request by RobertIndie**
Fixes #20420

### Motivation

When building the release candidate, the release mangers use their own account name to build and push the pulsar image. But currently the `pulsar-all` image always use `apachepulsar/pulsar` as the upstream image. And it may won't use the locally built image, which causes the issue.

### Modifications

* Specify the correct image name for `pulsar` image to build `pulsar-all`
* Remove the `latest` version from the build. The `latest` tag should only point to the latest major version. It's better to let the release manager specify the latest tag manually.
* Add github commit id to the image tag.

The image would like: apachepulsar/pulsar:3.1.0-SNAPSHOT-97a1a87

### Verifying this change

* Prepare the `pulsar` image: `docker tag apachepulsar/pulsar:3.0.0 snzkyang/pulsar:3.1.0-SNAPSHOT-97a1a87` 
* Run the build command of `pulsar-all` image:
```
mvn install \
        -DskipTests \
        -Pdocker \
        -Ddocker.platforms=linux/amd64 \
        -Ddocker.organization=snzkyang \
         -pl docker/pulsar-all
```
It will build successfully:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  09:29 min
[INFO] Finished at: 2023-05-30T11:29:28+08:00
[INFO] ------------------------------------------------------------------------
[INFO] 12 goals, 12 executed
```
* Verify
```
pulsar git:(iss-20420) ✗ docker run snzkyang/pulsar-all:3.1.0-SNAPSHOT-97a1a87 ls connectors
LICENSE
README
pulsar-io-aerospike-3.1.0-SNAPSHOT.nar
pulsar-io-alluxio-3.1.0-SNAPSHOT.nar
pulsar-io-batch-data-generator-3.1.0-SNAPSHOT.nar
pulsar-io-canal-3.1.0-SNAPSHOT.nar
pulsar-io-cassandra-3.1.0-SNAPSHOT.nar
pulsar-io-data-generator-3.1.0-SNAPSHOT.nar
pulsar-io-debezium-mongodb-3.1.0-SNAPSHOT.nar
...
```

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by lhotari** [comment]
@RobertIndie there's already #20433 which contains a better solution and on the way to preventing the issue.

**Comment by RobertIndie** [comment]
> @RobertIndie there's already #20433 which contains a better solution and on the way to preventing the issue. 

This PR is focusing on the fix. And it has been tested and verified. From my understanding, #20433 may have the problem when building the image and still need some changes. Please see https://github.com/apache/pulsar/pull/20433#pullrequestreview-1450186015 for more detail.

**Review by lhotari — CHANGES_REQUESTED** [review]
* Git SHA should be part of the tag (it could be an extra tag) so that it is unique and ensures that the correct parent image is used, also in the case that the build ordering wouldn't be effective for some reason.
  * what if "pulsar-all" gets built before "pulsar" in the build?
* Pushing the latest tag should be prevented in maintenance branch publishing process and scripts used to do the publishing

**Review by tisonkun — APPROVED** [review]

File: docker/pulsar/pom.xml [hunk 1]
```diff
@@ -57,6 +57,29 @@
       <id>docker</id>
       <build>
         <plugins>
+          <plugin>
+            <groupId>pl.project13.maven</groupId>
+            <artifactId>git-commit-id-plugin</artifactId>
```
Comment (tisonkun) [inline docker/pulsar/pom.xml:62]: Although, this plugin seems a bit wordy and it's licensed under LGPL (although we don't bundle its source code so it should not be an issue).

I can trust [mavanagaiata](https://github.com/koraktor/mavanagaiata) plugin that has been used in ZooKeeper projects for years.

File: docker/pulsar-all/pom.xml [hunk 2]
```diff
@@ -125,6 +125,29 @@
...
+              <useNativeGit>true</useNativeGit>
+              <prefix>git</prefix>
+              <failOnNoGitDirectory>true</failOnNoGitDirectory>
+              <skipPoms>false</skipPoms>
+              <gitDescribe>
+                <skip>false</skip>
+                <always>false</always>
+              </gitDescribe>
+            </configuration>
+          </plugin>
```
Comment (lhotari) [inline docker/pulsar-all/pom.xml:150]: would it be possible to apply this in the parent `pom.xml`? This plugin will be useful in other contexts too.

File: docker/pulsar/pom.xml [hunk 3]
```diff
@@ -81,8 +104,7 @@
                         </args>
                         <contextDir>${project.basedir}</contextDir>
                         <tags>
-                          <tag>latest</tag>
-                          <tag>${project.version}</tag>
```
Comment (lhotari) [inline docker/pulsar/pom.xml:85]: I don't think that we need to remove these tags. It could be useful to keep them. 

For consistency, the new tag should also be added to all other images that are built.
for example 
`docker/pulsar-all/pom.xml`
`tests/docker-images/latest-version-image/pom.xml`
`tests/docker-images/java-test-image/pom.xml`.

`tests/docker-images/latest-version-image/Dockerfile` should reference the parent image in a similar  as `docker/pulsar-all/Dockerfile`

**Review by lhotari — CHANGES_REQUESTED** [review]

**Comment by lhotari** [comment]
One aspect to check is what happens when the docker image is built from the sources .tar.gz file without a git repository. I'm not sure if that has worked in the past so I'm not sure if it needs to work. It would be useful to have it working.

File: docker/pulsar-all/pom.xml [hunk 4]
```diff
@@ -125,6 +125,29 @@
...
+                <id>git-info</id>
+                <goals>
+                  <goal>revision</goal>
+                </goals>
+              </execution>
+            </executions>
+            <configuration>
+              <useNativeGit>true</useNativeGit>
+              <prefix>git</prefix>
+              <failOnNoGitDirectory>true</failOnNoGitDirectory>
```
Comment (lhotari) [inline docker/pulsar-all/pom.xml:143]: I think that we would need to support building directly from ASF provided sources .tar.gz file. In that case there is no git directory.

Comment (tisonkun) [reply docker/pulsar-all/pom.xml:143]: Not always. We can add a profile to enable build only with Git - we're fixing a release issue now, not generally improve the build experience.

Comment (tisonkun) [reply docker/pulsar-all/pom.xml:143]: Support build without Git can be a follow-up. Either we calculate hash from files, required a passed tag, or keep using latest.

File: docker/pulsar/pom.xml (hunk 1 above)
Comment (RobertIndie) [reply docker/pulsar/pom.xml:62]: We already used this plugin in the codebase:
https://github.com/apache/pulsar/blob/43b3622cc7e7f746ca9920fdc704dc7448767ac7/pom.xml#L1928-L1930

I think we can consider replacing it with another PR.

Comment (tisonkun) [reply docker/pulsar/pom.xml:62]: .SGTM

File: docker/pulsar-all/pom.xml (hunk 2 above)
Comment (RobertIndie) [reply docker/pulsar-all/pom.xml]: Good idea. I moved it to the parent `pom.xml` and skip it by default for other modules.

File: docker/pulsar-all/pom.xml (hunk 4 above)
Comment (RobertIndie) [reply docker/pulsar-all/pom.xml]: If you don't set it to false, it will be unable to resolve `${git.commit.id.abbrev}`.

But I use a simple way to enable it to build the image without the git directory. The image name will be like `apachepulsar/pulsar:3.1.0-no-git`. We can further improve it with another PR if you have a better idea.

File: docker/pulsar/pom.xml (hunk 3 above)
Comment (RobertIndie) [reply docker/pulsar/pom.xml]: That makes sense to me. I added the `latest` tag back.

**Comment by RobertIndie** [comment]
> One aspect to check is what happens when the docker image is built from the sources .tar.gz file without a git repository. I'm not sure if that has worked in the past so I'm not sure if it needs to work. It would be useful to have it working.

@lhotari I have pushed a [commit](https://github.com/apache/pulsar/pull/20435/commits/4bd97d7cc071687db0007e4a5a34c56019549a0c) to enable building the image without git repo. And I have verified it works. The image name will be like `snzkyang/pulsar:3.1.0-SNAPSHOT-no-git` if there is no git directory.

File: pom.xml [hunk 5]
```diff
@@ -1571,6 +1571,31 @@ flexible messaging model and an intuitive client API.</description>
...
+        <executions>
+          <execution>
+            <id>git-info</id>
+            <goals>
+              <goal>revision</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <skip>true</skip>
```
Comment (lhotari) [inline pom.xml:1587]: Just thinking whether `pulsar-common`'s usage of `git-commit-id-plugin` continues to work as expected. It should since by default all configuration is replaced if it's defined in a child pom.xml?

File: docker/pulsar/pom.xml [hunk 6]
```diff
@@ -48,15 +48,23 @@
...
       <id>docker</id>
       <build>
         <plugins>
+          <plugin>
+            <groupId>pl.project13.maven</groupId>
+            <artifactId>git-commit-id-plugin</artifactId>
+            <configuration>
+              <skip>false</skip>
+            </configuration>
+          </plugin>
```
Comment (lhotari) [inline docker/pulsar/pom.xml:67]: I think that this should be defined in `docker/pom.xml` so that it gets applied to both `docker/pulsar/pom.xml` and `docker/pulsar-all/pom.xml`.

**Review by lhotari — CHANGES_REQUESTED** [review]
